### DISCUSSION TITLE
Using `spdlog::level::off` to unconditionally log a message, OK?

### DISCUSSION THREAD

**Discussion by N-Dekker** [gabime/spdlog#2640] [Q&A]
I would like to replace a few print statements in an [existing code base](https://github.com/SuperElastix/elastix) by corresponding spdlog calls. Those old print statements were _unconditional_ (independent of any logging level), but not necessarily "critical". Is that a valid use case for `spdlog::level::off`? For example: 

     my_logger.log(spdlog::level::off, message);

Just to be sure!  😃

**Comment by tt4g** [comment]
`spdlog::level::off` is used to suppress all logging. Using `spdlog::level::trace` will output all logs.

https://github.com/gabime/spdlog/blob/927cc29444a294d76e83dfb898e797dc431ce094/include/spdlog/logger.h#L300-L304

https://github.com/gabime/spdlog/blob/927cc29444a294d76e83dfb898e797dc431ce094/include/spdlog/common.h#L223-L235

**Reply by N-Dekker** [reply]
OK, thanks, but I would like some particular messages to _always_ appear. _Independent_ of the log level of the sinks (or the level of the logger). Is that possible?

**Reply by tt4g** [reply]
Log level filters cannot be ignored.

Temporarily change the log level to `spdlog::level::trace` or use `print()`.
Alternatively, use a dedicated logger and sink (both at log level `spdlog::level::trace`) for logs you want to always output.

**Reply by N-Dekker** [reply]
Thanks, but looking at https://github.com/gabime/spdlog/blob/927cc29444a294d76e83dfb898e797dc431ce094/include/spdlog/logger.h#L300-L304

It seems to me that when `msg_level` is high enough, `logger::should_log(msg_level)` will return true for any `logger::level_`.

**Reply by tt4g** [reply]
Sorry, I realized my mistake.
You are thinking of setting the log level `spdlog::level::off` for log messages.
Since `spdlog::logger::off()` does not exist, I assumed it was the filter level of the logger.

If the level that the message has is `spdlog::level::off`, as you first guessed, then the log is not filtered.

**Reply by N-Dekker** [reply]
Thanks @tt4g My impression was already that `my_logger.log(spdlog::level::off, message)` just works well, when this particular message needs to be printed unconditionally, unfiltered. I just wanted to be sure that it is considered OK to do so. So that it won't trigger any assert failure, abort, exception or undefined behavior from spdlog itself!

**Reply by tt4g** [reply]
Undefined behavior should not occur in such cases. `spdlog::level::off` should be treated by the code the same way as `spdlog::level::info` when it is output.

One thing to note is that `flush()` is not triggered:

https://github.com/gabime/spdlog/blob/8512000f36c2ad9b1265bd78b11c0b34151d6be4/include/spdlog/logger-inl.h#L221-L225

**Reply by N-Dekker** [reply]
Thanks @tt4g I did not realize before that for a message level, the value `level::off` suppressed automatic flushing. But I guess an explicit  `logger::flush()` call in user code still flushes such "level::off messages", right?

**Reply by tt4g** [reply]
`logger::flush()` always works. Automatic flushing after log message output only is not performed.

**Reply by N-Dekker** [reply]
So I learned here that when a specific message itself has message log level `off`: 

1. such a message is logged unconditionally, unfiltered, independent of any logger's log level
2. such a message is _not_ flushed automatically 

Has this specific behavior of logging a "level off message" already been documented? If not, do you think this could still be added, for example to the [wiki](https://github.com/gabime/spdlog/wiki)?

**Reply by tt4g** [reply]
There is no documentation for `spdlog::level::off` as far as I know (there is a section on filtering by log level in `README.md`).
Since the wiki is community based and maintained, you can add an explanation when setting `spdlog::level::off` to the log message.

**Reply by tt4g** [reply]
Added: https://github.com/gabime/spdlog/wiki/0.-FAQ#how-to-force-log-messages-to-be-output-ignore-log-level-filter

I am not a native speaker, so please edit if there is anything that needs to be edited.
